File: pequenas_coisas/XMLParsingTest.java
import org.junit.*;
import static org.junit.Assert.*;
import org.hamcrest.*;
import static org.hamcrest.CoreMatchers.*;
import javax.xml.parsers.*;
import org.w3c.dom.*;
import java.io.*;

public class XMLParsingTest {
    @Test public void getTheCorrectNumberOfPeopleOnXML() throws Exception {
        DocumentBuilderFactory docBuilderF = DocumentBuilderFactory.newInstance();
        DocumentBuilder docBuilder = docBuilderF.newDocumentBuilder();
        Document doc = docBuilder.parse(new ByteArrayInputStream(xml().getBytes()));
        assertThat(doc.getElementsByTagName("person").getLength(), is(2));
    }

    @Test public void getTheCorrectNumberOfPeople() throws Exception {
        assertThat(xml(), have(2).tags("person"));
    }

    private TagMatcher have(int count) {
        return new TagMatcher(count);
    }

    class TagMatcher {
        private int count;
        public TagMatcher(int count) { this.count = count; }

        public Matcher<String> tags(final String tagName) {
            return new BaseMatcher<String>() {
                public void describeTo(Description d) { 
                    d.appendText("to have " + count + " <" + tagName + "> tags"); 
                }

                public boolean matches(Object xmlObj) {
                    try {
                        String xml = (String) xmlObj;
                        DocumentBuilderFactory docBuilderF = DocumentBuilderFactory.newInstance();
                        DocumentBuilder docBuilder = docBuilderF.newDocumentBuilder();
                        Document doc = docBuilder.parse(new ByteArrayInputStream(xml.getBytes()));
                        return doc.getElementsByTagName(tagName).getLength() == count;
                    } catch(Exception e) {
                        return false;
                    }
                }
            };
        }

    }

    private String xml() {
      return "<people><person>Ariovaldo Steil</person>" +
        "<person>Ariovaldo Gumercindo</person></people>";
    }
}

File: pequenas_coisas/XMLParsingTest.scala
import org.scalatest.FlatSpec
import org.scalatest.matchers._
import scala.xml._

class XMLParsingTest extends FlatSpec with ShouldMatchers {
    it should "get the correct number of people on XML" in {
        (xml \ "person").size should be === (2)
    }

    it should "get the correct number of people" in {
        xml should have(2).tags("person")
    }

    def xml = <people><person>Ariovaldo Steil</person><person>Ariovaldo Gumercindo</person></people>

    def have(count: Int) = new TagMatcher(count)
    
    class TagMatcher(count: Int) {
        def tags(tag: String) = new Matcher[Seq[Node]] {
            def apply(xml: Seq[Node]) = MatchResult(
                (xml \ tag).size == count,
                "expected to have " + count + " <" + tag + "> tags",
                "expected to not have " + count + " <" + tag + "> tags"
            )
        }
    }
}
